File: 3.java
import java.util.*;

public class AStar8Puzzle {

    // Print the state
    static void printState(int[][] state) {
        for (int[] row : state) {
            for (int val : row) {
                System.out.print(val + " ");
            }
            System.out.println();
        }
    }

    // Find blank (0)
    static int[] findBlank(int[][] state) {
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (state[i][j] == 0) {
                    return new int[]{i, j};
                }
            }
        }
        return null;
    }

    static int[][] copyState(int[][] state) {
        int[][] newState = new int[3][3];
        for (int i = 0; i < 3; i++) {
            newState[i] = state[i].clone();
        }
        return newState;
    }

    static int[][] moveUp(int[][] state) {
        int[] pos = findBlank(state);
        int i = pos[0], j = pos[1];
        if (i > 0) {
            int[][] newState = copyState(state);
            newState[i][j] = newState[i - 1][j];
            newState[i - 1][j] = 0;
            return newState;
        }
        return null;
    }

    static int[][] moveDown(int[][] state) {
        int[] pos = findBlank(state);
        int i = pos[0], j = pos[1];
        if (i < 2) {
            int[][] newState = copyState(state);
            newState[i][j] = newState[i + 1][j];
            newState[i + 1][j] = 0;
            return newState;
        }
        return null;
    }

    static int[][] moveLeft(int[][] state) {
        int[] pos = findBlank(state);
        int i = pos[0], j = pos[1];
        if (j > 0) {
            int[][] newState = copyState(state);
            newState[i][j] = newState[i][j - 1];
            newState[i][j - 1] = 0;
            return newState;
        }
        return null;
    }

    static int[][] moveRight(int[][] state) {
        int[] pos = findBlank(state);
        int i = pos[0], j = pos[1];
        if (j < 2) {
            int[][] newState = copyState(state);
            newState[i][j] = newState[i][j + 1];
            newState[i][j + 1] = 0;
            return newState;
        }
        return null;
    }

    // Heuristic: number of misplaced tiles
    static int calculateHeuristic(int[][] state, int[][] goal) {
        int h = 0;
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (state[i][j] != goal[i][j]) {
                    h++;
                }
            }
        }
        return h;
    }

    static String stateToString(int[][] state) {
        StringBuilder sb = new StringBuilder();
        for (int[] row : state) {
            for (int val : row) {
                sb.append(val);
            }
        }
        return sb.toString();
    }

    static void aStar(int[][] initialState, int[][] goalState) {

        // OPEN: (f, g, state)
        List<Node> OPEN = new ArrayList<>();
        Set<String> CLOSED = new HashSet<>();

        int h0 = calculateHeuristic(initialState, goalState);
        OPEN.add(new Node(h0, 0, initialState));

        while (!OPEN.isEmpty()) {

            // Get node with minimum f
            Node current = Collections.min(OPEN, Comparator.comparingInt(n -> n.f));
            OPEN.remove(current);

            CLOSED.add(stateToString(current.state));

            printState(current.state);
            System.out.println();

            if (Arrays.deepEquals(current.state, goalState)) {
                System.out.println("Solution found!");
                return;
            }

            int[][][] moves = {
                    moveUp(current.state),
                    moveDown(current.state),
                    moveLeft(current.state),
                    moveRight(current.state)
            };

            for (int[][] successor : moves) {
                if (successor == null) continue;

                String key = stateToString(successor);
                if (CLOSED.contains(key)) continue;

                int gNew = current.g + 1;
                int hNew = calculateHeuristic(successor, goalState);
                int fNew = gNew + hNew;

                OPEN.add(new Node(fNew, gNew, successor));
            }
        }

        System.out.println("No solution found.");
    }

    // Node class
    static class Node {
        int f, g;
        int[][] state;

        Node(int f, int g, int[][] state) {
            this.f = f;
            this.g = g;
            this.state = state;
        }
    }

    // Main method
    public static void main(String[] args) {

        int[][] initialState = {
                {1, 2, 3},
                {8, 0, 4},
                {7, 6, 5}
        };

        int[][] goalState = {
                {2, 8, 1},
                {0, 4, 3},
                {7, 6, 5}
        };

        aStar(initialState, goalState);
    }
}

File: 3.py

      
def print_state(state):
    for row in state:
        print(" ".join(map(str, row)))


def find_blank(state):
    for i in range(3):
        for j in range(3):
            if state[i][j] == 0:
                return i, j


def move_up(state):
    i, j = find_blank(state)
    if i > 0:
        new_state = [row[:] for row in state]
        new_state[i][j], new_state[i-1][j] = new_state[i-1][j], new_state[i][j]
        return new_state
    return None


def move_down(state):
    i, j = find_blank(state)
    if i < 2:
        new_state = [row[:] for row in state]
        new_state[i][j], new_state[i+1][j] = new_state[i+1][j], new_state[i][j]
        return new_state
    return None


def move_left(state):
    i, j = find_blank(state)
    if j > 0:
        new_state = [row[:] for row in state]
        new_state[i][j], new_state[i][j-1] = new_state[i][j-1], new_state[i][j]
        return new_state
    return None


def move_right(state):
    i, j = find_blank(state)
    if j < 2:
        new_state = [row[:] for row in state]
        new_state[i][j], new_state[i][j+1] = new_state[i][j+1], new_state[i][j]
        return new_state
    return None


def calculate_heuristic(state, goal_state):
    h = 0
    for i in range(3):
        for j in range(3):
            if state[i][j] != 0 and state[i][j] != goal_state[i][j]:
                h += 1
    return h


def a_star(initial_state, goal_state):
    OPEN = []
    CLOSED = set()

    h0 = calculate_heuristic(initial_state, goal_state)
    OPEN.append((h0, 0, initial_state))  # (f, g, state)

    while OPEN:
        # pick node with minimum f
        f, g, current_state = min(OPEN, key=lambda x: x[0])
        OPEN.remove((f, g, current_state))

        state_key = tuple(map(tuple, current_state))
        CLOSED.add(state_key)

        print_state(current_state)
        print()

        if current_state == goal_state:
            print("Solution found!")
            return

        successors = [
            move_up(current_state),
            move_down(current_state),
            move_left(current_state),
            move_right(current_state)
        ]

        for successor in successors:
            if successor is None:
                continue

            succ_key = tuple(map(tuple, successor))
            if succ_key in CLOSED:
                continue

            g_succ = g + 1
            h_succ = calculate_heuristic(successor, goal_state)
            f_succ = g_succ + h_succ

            # check if better path exists in OPEN
            in_open = False
            for f_old, g_old, state_old in OPEN:
                if state_old == successor and g_old <= g_succ:
                    in_open = True
                    break

            if not in_open:
                OPEN.append((f_succ, g_succ, successor))

    print("No solution found.")